astropy/utils/decorators.py

```diff
@@ -863,7 +866,212 @@ def __get__(self, obj, objtype=None):
+
+    def __post_init__(self):
+        object.__setattr__(self, "__doc__", getattr(self.fget, "__doc__", self.__doc__))
+        object.__setattr__(self, "exclude", set(self.exclude))
+
+    def _overriden__dir__(_self, self: object) -> List[str]:
+        """`object.__dir__` with filtering of entries."""
+        members = set(object.__dir__(self) if _self.fget is None else _self.fget(self))
+        members -= _self.exclude
+        return sorted(members)
+
+    # -------------------------------------------
+    # Descriptor
+
+    def __get__(self, encl_inst: Optional[object], encl_cls: Optional[type]=None) -> MethodType:
+        """Get ``override__dir__`` as a method of the enclosing instance/class."""
+        instance = encl_inst if encl_inst is not None else encl_cls
+        return MethodType(self._overriden__dir__, instance)
+
+    def getter(self, fget: MethodType) -> "_dir_":
+        return dataclasses.replace(self, fget=fget)
+
+    __call__ = getter
+
+
+class on_metaclass(classmethod):
+    """Decorate a method to make it emulate being defined on the metaclass.
+
+    Examples
+    --------
+    >>> class ExampleMeta(type):
+    ...     def identify(cls):
+    ...         print('this implements the {} method'.format(cls))
+    ...
+    >>> class Example(metaclass=ExampleMeta):
+    ...     @on_metaclass
+    ...     def identify(cls):
+    ...         print('this overrides the metaclass method for {}'.format(cls))
+    ...
+    >>> Example.identify()
+    this overrides the metaclass method for <class 'astropy.utils.decorators.Example'>
+
+    To reach the method on the metaclass we can do
+
+    >>> type(Example).identify(Example)
+    this implements the <class 'astropy.utils.decorators.Example'> method
+
+    Just like normal metaclass methods, ``identify`` cannot be called from an
+    instance of ``Example``.
+
+    >>> try:
+    ...     Example().identify()
+    ... except AttributeError as e:
+    ...     print(e)
+    'Example' object has no attribute 'identify'
+
+    Likewise, ``identify`` is not in the :func:`dir` of instances of
+    ``Example``.
+
+    >>> print("identify" in dir(Example()))
+    False
+
+    `on_metaclass` can also be used to promote a `classmethod` to a
+    `classmethod` on the metaclass, making a metaclass-classmethod.
+
+    >>> class Example(metaclass=ExampleMeta):
+    ...     @on_metaclass
+    ...     @classmethod
+    ...     def identify(mcls):
+    ...         print('this overrides the metaclass method for {}'.format(mcls))
+    ...
+    >>> Example.identify()
+    this overrides the metaclass method for <class 'astropy.utils.decorators.ExampleMeta'>
+
+    `on_metaclass` also works with `property` to make a class-level property.
+
+    >>> class Example(metaclass=ExampleMeta):
+    ...     @on_metaclass
+    ...     @property
+    ...     def identify(cls):
+    ...         print('this overrides the metaclass method for {}'.format(cls))
```

Comment:
Shouldn't a `property` return instead of printing?
```suggestion
    ...     @property
    ...     def identify(cls):
    ...         return f'this overrides the metaclass method for {cls}'
```